// phaser-list-view/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ScrollerEventDispatcher = exports.BasicSwiper = exports.DirectionalScroller = exports.WheelScroller = exports.SwipeCarousel = exports.ListViewCore = exports.ListView = exports.Scroller = undefined;

var _scroller = require('./scroller');

var _scroller2 = _interopRequireDefault(_scroller);

var _list_view = require('./list_view');

var _list_view2 = _interopRequireDefault(_list_view);

var _list_view_core = require('./list_view_core');

var _list_view_core2 = _interopRequireDefault(_list_view_core);

var _swipe_carousel = require('./swipe_carousel');

var _swipe_carousel2 = _interopRequireDefault(_swipe_carousel);

var _wheel_scroller = require('./wheel_scroller');

var _wheel_scroller2 = _interopRequireDefault(_wheel_scroller);

var _directional_scroller = require('./directional_scroller');

var _directional_scroller2 = _interopRequireDefault(_directional_scroller);

var _basic_swiper = require('./basic_swiper');

var _basic_swiper2 = _interopRequireDefault(_basic_swiper);

var _scroller_event_dispatcher = require('./scroller_event_dispatcher');

var _scroller_event_dispatcher2 = _interopRequireDefault(_scroller_event_dispatcher);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var PhaserListView = {
  Scroller: _scroller2.default,
  ListView: _list_view2.default,
  ListViewCore: _list_view_core2.default,
  SwipeCarousel: _swipe_carousel2.default,
  WheelScroller: _wheel_scroller2.default,
  DirectionalScroller: _directional_scroller2.default,
  BasicSwiper: _basic_swiper2.default,
  ScrollerEventDispatcher: _scroller_event_dispatcher2.default
};

exports.Scroller = _scroller2.default;
exports.ListView = _list_view2.default;
exports.ListViewCore = _list_view_core2.default;
exports.SwipeCarousel = _swipe_carousel2.default;
exports.WheelScroller = _wheel_scroller2.default;
exports.DirectionalScroller = _directional_scroller2.default;
exports.BasicSwiper = _basic_swiper2.default;
exports.ScrollerEventDispatcher = _scroller_event_dispatcher2.default;

// NOTE: we should only attach to the window in a production build

window.PhaserListView = PhaserListView;

exports.default = PhaserListView;
